### PR TITLE
Print jupyter notebook token when starting web UI.

### PR DESCRIPTION
**Pull request by Wapaul1**
This helps to address #870. Now the user just needs to get access to the port the server is launched on, e.g. by port forwarding.

### CODE REVIEW

**Comment by robertnishihara** [comment]
One concern I have is that this seems to slow down the `ray.init()` call a decent amount (e.g., by 2 or 3 seconds).

File: python/ray/worker.py [hunk 1]
```diff
@@ -888,6 +888,17 @@ def get_gpu_ids():
     """
     return global_worker.local_scheduler_client.gpu_ids()
 
+def webui_url_helper(client):
```
Comment (robertnishihara) [inline python/ray/worker.py]: Why do we need a helper function?

**Comment by robertnishihara** [comment]
The extra time required by `ray.init()` is a pretty big deal. Let's look into the following.
- can we provide our own token so that we don't have to wait to figure it out
- how does setting a notebook password work, does that make sense

**Comment by Wapaul1** [comment]
We now generate a token and set the server with it, so there shouldn't be an added delay to `ray.init`.

File: python/ray/__init__.py [hunk 2]
```diff
@@ -12,7 +12,7 @@
 
 from ray.worker import (register_class, error_info, init, connect, disconnect,
                         get, put, wait, remote, log_event, log_span,
-                        flush_log, get_gpu_ids)  # noqa: E402
+                        flush_log, get_gpu_ids, get_webui_url)  # noqa: E402
```
Comment (robertnishihara) [inline python/ray/__init__.py]: Linting says this is imported but unused. The way to deal with this is to add it to the list `__all__` below.

File: python/ray/services.py [hunk 3]
```diff
@@ -492,10 +493,12 @@ def start_ui(redis_address, stdout_file=None, stderr_file=None, cleanup=True):
             port += 1
     new_env = os.environ.copy()
     new_env["REDIS_ADDRESS"] = redis_address
+    token = binascii.hexlify(os.urandom(24))
```
Comment (robertnishihara) [inline python/ray/services.py]: Add a comment explaining why we are generating our own token and what the token will be used for.

File: python/ray/worker.py [hunk 4]
```diff
@@ -1681,6 +1696,8 @@ def connect(info, object_id_seed=None, mode=WORKER_MODE, worker=global_worker,
         driver_info["name"] = (main.__file__ if hasattr(main, "__file__")
                                else "INTERACTIVE MODE")
         worker.redis_client.hmset(b"Drivers:" + worker.worker_id, driver_info)
+        if not worker.redis_client.exists("webui"):
+            worker.redis_client.hmset("webui", {"url": info["webui_url"]})
```
Comment (robertnishihara) [inline python/ray/worker.py]: Should we use `set` or do you think we'll want to add more fields down the road?

**Review by robertnishihara — COMMENTED** [review]
This looks good, can you address my comments and fix the linting errors? (also look into the MacOS test failures)

File: python/ray/services.py [hunk 5]
```diff
@@ -14,6 +14,7 @@
 import sys
 import time
 import threading
+import binascii
```
Comment (robertnishihara) [inline python/ray/services.py]: These should be in alphabetical order.

File: python/ray/worker.py [hunk 6]
```diff
@@ -1088,7 +1101,9 @@ def get_address_info_from_redis_helper(redis_address, node_ip_address):
     client_info = {"node_ip_address": node_ip_address,
                    "redis_address": redis_address,
                    "object_store_addresses": object_store_addresses,
-                   "local_scheduler_socket_names": scheduler_names}
+                   "local_scheduler_socket_names": scheduler_names,
+                   # Web UI should be running.
+                   'webui_url': webui_url_helper(redis_client)}
```
Comment (robertnishihara) [inline python/ray/worker.py]: Please be consistent with the surrounding code, so if all of the keys are strings using double quotes then when you add a new key it should be formatted the same way.

File: python/ray/worker.py [hunk 7]
```diff
@@ -889,6 +889,19 @@ def get_gpu_ids():
     return global_worker.local_scheduler_client.gpu_ids()
 
 
+def webui_url_helper(client):
```
Comment (robertnishihara) [inline python/ray/worker.py]: This should be prefixed with an underscore

File: python/ray/worker.py [hunk 8]
```diff
@@ -889,6 +889,19 @@ def get_gpu_ids():
...
+    """Parsing for getting the url of the web UI."""
+    return client.hmget('webui', 'url')[0]
```
Comment (robertnishihara) [inline python/ray/worker.py]: In python 3 this returns a bytes object and it should return a string

**Comment by robertnishihara** [comment]
@Wapaul1 I changed the print back to 70 characters instead of 123 because terminals are often 80 characters and the text ended up wrapping.

**Pull request merged**